typhon/engine.py

```diff
@@ -0,0 +1,227 @@
+############
+# Standard #
+############
+import logging
+
+
+############
+# External #
+############
+from bluesky import RunEngine
+from bluesky.utils import RunEngineInterrupted, install_qt_kicker
+from pydm.PyQt.QtGui import QVBoxLayout, QLabel, QComboBox, QGroupBox
+from pydm.PyQt.QtGui import QStackedWidget, QPushButton
+from pydm.PyQt.QtCore import QObject, pyqtSignal, pyqtSlot
+
+
+logger = logging.getLogger(__name__)
+
+
+class QRunEngine(QObject, RunEngine):
+
+    state_changed = pyqtSignal('QString', 'QString')
+    update_rate = 0.02
+    command_registry = {'Halt': RunEngine.halt,
+                        'Abort': RunEngine.abort,
+                        'Resume': RunEngine.resume,
+                        'Pause': RunEngine.request_pause,
+                        'Stop': RunEngine.stop}
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Attach the state_hook to emit signals
+        self.state_hook = self.on_state_change
+        # Create a kicker, not worried about doing this multiple times as this
+        # is checked by `install_qt_kicker` itself
+        install_qt_kicker(update_rate=self.update_rate)
+        # Allow a plan to be stored on the RunEngine
+        self.plan_creator = None
+
+    def on_state_change(self, state, old_state):
+        """
+        Report a state change of the RunEngine
+
+        This is added directly to the `RunEngine.state_hook` and emits the
+        `engine_state_change` signal.
+
+        Parameters
+        ----------
+        state: str
+
+        old_state: str
+
+        """
+        self.state_changed.emit(state, old_state)
+
+    @pyqtSlot()
+    def start(self):
```

Comment:
maybe require that the start slot take in a plan? That moves plan selection logic out to the containing widget and lets you do things like a drop-down list, full UIs for plans or a text box to just type the plan in.

Reply:
👍 to this idea. This simplifies the concerns about attaching plan. It will require a little more work for users though. It probably means most applications will have to have do something like:

```python
but  = QPushButton('My Plan')
but.clicked.connect(functools.partial(QRE.start, my_plan()))
```

or in your example look up the plan from a dictionary based on a string. This is maybe slightly awkward but a lot more flexible. I can always wrap this up in a "PlanButton" or something. Also is there a point to having this by a `pyqtSlot`, can't imagine a lot of buttons will be emitting plans. We probably want to enforce this is a one to one relationship.

Reply:
plans are all standard python objects, I don't see too much trouble passing live generators across a slot (the c++ layer should just pass it along as an opaque object).